keras_nlp/layers/modeling/rotary_embedding.py

```diff
@@ -16,6 +16,40 @@
 
 
 class RotaryEmbedding(keras.layers.Layer):
+    """Rotary positional encoding layer.
+
+    This layer encodes absolute positional information with rotation
+    matrix. It calculates the rotary encoding with a mix of sine and
+    cosine functions with geometrically increasing wavelengths.
+    Defined and formulated in [RoFormer: Enhanced Transformer with Rotary Position Embedding](https://arxiv.org/abs/2104.09864v4).
+    Takes as input the query and key tensors. The input must have shape
+    [batch_size, num_heads, sequence_length, query_length]. This layer
+    will return new query and key tensors after applying rotational
+    encoding.
+
+    Args:
+        percentage: float. The percentage of attn_head_size over which rotation
+            should be applied. Defaults to 0
+        max_wavelength: int. The maximum angular wavelength of the sine/cosine
+            curves, as described in Attention is All You Need. Defaults to
+            `10000`.
+
+    Examples:
+
+    ```python
+    batch_size = 16
+    num_heads = 8
+    sequence_length = 256
+    query_length = 256
+    query = tf.ones((batch_size, num_heads, sequence_length, query_length))
+    rot_emb_layer = RotaryEmbedding()
+    query_rot = rot_emb_layer(query)
+    ```
+
+    References:
+     - [RoFormer: Enhanced Transformer with Rotary Position Embedding](https://arxiv.org/abs/2104.09864v4)
+    """
+
     def __init__(self, max_wavelength=10000, **kwargs):
         super().__init__(**kwargs)
         self.max_wavelength = max_wavelength
```

Comment:
A few comments for below (github won't let me comment). Why is the following necessary?

```
cos_emb = cos_emb[:, : ops.shape(tensor)[1], :, :]
sin_emb = sin_emb[:, : ops.shape(tensor)[1], :, :]
```

The cos/sin embeddings should already have `seq_len` shape when you compute them.

Lastly, if you wanted to make `_compute_cos_sin_embedding` work with any number of dimensions, you would need to update it. Here's a draft of a change, but haven't tested this yet.

```
embedding = ops.concatenate((freqs, freqs), axis=-1)
for dim in range(len(x.shape)):
    if axis != self.sequence_axis and axis != self.feature_axis:
        embedding = ops.expand_dims(embedding, axis)
```

Reply:
Thanks for the pointer, got it working with some tweaks !